File: academia-profesional/proyecto-final-erp/frontend/src/app/core/services/producto.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Producto {
  id: number; nombre: string; descripcion: string; precio: number;
  costo: number; stock: number; sku: string; categoriaId: number; activo: boolean;
}

@Injectable({ providedIn: 'root' })
export class ProductoService {
  private http = inject(HttpClient);
  private api = '/api/productos';

  listar(page = 0, size = 20): Observable<any> { return this.http.get(`${this.api}?page=${page}&size=${size}`); }
  buscar(q: string): Observable<Producto[]> { return this.http.get<Producto[]>(`${this.api}/buscar?q=${q}`); }
  stockBajo(min = 10): Observable<Producto[]> { return this.http.get<Producto[]>(`${this.api}/stock-bajo?min=${min}`); }
  crear(p: Partial<Producto>): Observable<Producto> { return this.http.post<Producto>(this.api, p); }
  actualizar(id: number, p: Partial<Producto>): Observable<Producto> { return this.http.put<Producto>(`${this.api}/${id}`, p); }
  eliminar(id: number): Observable<void> { return this.http.delete<void>(`${this.api}/${id}`); }
}

File: academia-profesional/proyecto-07-api-productos/src/main/java/com/academia/productos/service/ProductoService.java
package com.academia.productos.service;

import com.academia.productos.dto.ProductoRequest;
import com.academia.productos.dto.ProductoResponse;
import com.academia.productos.entity.Producto;
import com.academia.productos.repository.ProductoRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional(readOnly = true)
public class ProductoService {

    private final ProductoRepository repo;

    public ProductoService(ProductoRepository repo) { this.repo = repo; }

    public Page<ProductoResponse> listar(Pageable pageable) {
        return repo.findByActivoTrue(pageable).map(this::toResponse);
    }

    public ProductoResponse buscarPorId(Long id) {
        return repo.findById(id).map(this::toResponse)
                .orElseThrow(() -> new RuntimeException("Producto no encontrado: " + id));
    }

    public List<ProductoResponse> buscar(String nombre) {
        return repo.findByNombreContainingIgnoreCase(nombre).stream().map(this::toResponse).toList();
    }

    public List<ProductoResponse> porCategoria(String categoria) {
        return repo.findByCategoria(categoria).stream().map(this::toResponse).toList();
    }

    public List<String> categorias() { return repo.findCategorias(); }

    public List<ProductoResponse> stockBajo(int minimo) {
        return repo.findStockBajo(minimo).stream().map(this::toResponse).toList();
    }

    @Transactional
    public ProductoResponse crear(ProductoRequest req) {
        var p = new Producto();
        p.setNombre(req.nombre());
        p.setPrecio(req.precio());
        p.setCosto(req.costo());
        p.setStock(req.stock());
        p.setSku(req.sku());
        p.setCategoria(req.categoria());
        p.setDescripcion(req.descripcion());
        return toResponse(repo.save(p));
    }

    @Transactional
    public ProductoResponse actualizar(Long id, ProductoRequest req) {
        var p = repo.findById(id).orElseThrow(() -> new RuntimeException("No encontrado: " + id));
        if (req.nombre() != null) p.setNombre(req.nombre());
        if (req.precio() != null) p.setPrecio(req.precio());
        if (req.descripcion() != null) p.setDescripcion(req.descripcion());
        if (req.categoria() != null) p.setCategoria(req.categoria());
        p.setStock(req.stock());
        return toResponse(repo.save(p));
    }

    @Transactional
    public void eliminar(Long id) {
        var p = repo.findById(id).orElseThrow(() -> new RuntimeException("No encontrado: " + id));
        p.setActivo(false);
        repo.save(p);
    }

    private ProductoResponse toResponse(Producto p) {
        return new ProductoResponse(p.getId(), p.getNombre(), p.getDescripcion(),
                p.getPrecio(), p.getCosto(), p.getStock(), p.getSku(), p.getCategoria(), p.getCreatedAt());
    }
}
